stdlib/dynarray.mli

```diff
@@ -0,0 +1,418 @@
+    [1], [2], and [3] at the end of [a].
+    [append_iter a Queue.iter q] adds elements from the queue [q]. *)
+
+
+(** {1:removing Removing elements} *)
+
+val pop_last_opt : 'a t -> 'a option
+(** [pop_last_opt a] removes and returns the last element of [a],
+    or [None] if the array is empty. *)
+
+val pop_last : 'a t -> 'a
+(** [pop_last a] removes and returns the last element of [a].
+
+    @raise Not_found on an empty array. *)
+
+val remove_last : 'a t -> unit
+(** [remove_last a] removes the last element of [a] , or does nothing
+    if [is_empty a].
+*)
+
+val clear : 'a t -> unit
+(** [clear a] removes all the elements of [a].
+
+    It is equivalent to [truncate a 0].
+
+    Similar to {!Buffer.clear}.
+*)
+
+val truncate : 'a t -> int -> unit
+(** [truncate a n] truncates [a] to have at most [n] elements.
+
+    It removes elements whose index is great or equal than [n].
+    It does nothing if [n >= length a].
+
+    It is equivalent to:
+    {[
+      while length a > n do
+        remove_last a
+      done
+    ]}
+
+    Similar to {!Buffer.truncate}.
+*)
+
+
+(** {1:iteration Iteration}
+
+    The iteration functions traverse the elements of a dynamic array.
+
+    It is a programming error to modify the length of an array
+    (by adding or removing elements) during an iteration on the
+    array. Any iteration function will fail with [Invalid_argument]
+    if it detects such a length change.
+*)
+
+val iter : ('a -> unit) -> 'a t -> unit
+(** [iter f a] calls [f] on each element of [a], in increasing index order. *)
+
+val iteri : (int -> 'a -> unit) -> 'a t -> unit
+(** [iteri f a] calls [f i x] for each [x] at index [i] in [a]. *)
+
+val map : ('a -> 'b) -> 'a t -> 'b t
+(** [map f a] is a new array of length [length a], with elements mapped
+    from [a] using [f]. *)
+
+val mapi : (int -> 'a -> 'b) -> 'a t -> 'b t
+(** [mapi f v] is just like {!map}, but it also passes in the index
+    of each element as the first argument to the function [f]. *)
+
+val fold_left : ('acc -> 'a -> 'acc) -> 'acc -> 'a t -> 'acc
+(** [fold_left f acc a] folds [f] over [a] starting with accumulator [acc]. *)
+
+val exists : ('a -> bool) -> 'a t -> bool
+(** [exists f a] returns [true] if some element of [a] satisfies [f]. *)
+
+val for_all : ('a -> bool) -> 'a t -> bool
+(** [for_all f a] returns [true] if all elements of [a] satisfie [f].
+    This includes the case where [a] is empty. *)
+
+val filter : ('a -> bool) -> 'a t -> 'a t
+(** [filter f a] is an array containing all elements of [a] that satisfy [f] *)
+
+val filter_map : ('a -> 'b option) -> 'a t -> 'b t
+(** [filter_map f a] is a new array [b], such that for each item [x] in [a]:
+  - if [f x = Some y], then [y] is in [b]
+  - if [f x = None], then no element is added to [b]. *)
+
+
+(** {1:conversions Conversions to other data structures}
+
+    Note: the [of_*] functions can raise [Failure] if the length would
+    need to grow beyond {!Sys.max_array_length}.
```

Comment:
Can I push back on the use of `Invalid_argument` here? If we assume that `Failure` is for catchable failures due to unsupported inputs, while `Invalid_argument` is for programming errors due to out-of-spec input, then let's consider the following reimplementation of `init`

```ocaml
let init n f = of_list (List.init n f)
```

Is this function correct (but partially defined) or buggy because it is a programming error to not check the `List.init` precondition? I would say "corect", and I think that raising a `Failure` here is the better choice.

I get that `Array.of_list` behaves differently, but:
- I would consider this a mistake coming from a time where the difference between the two errors was not clear-cut
- If people respect our guidelines to not catch `Invalid_argument` explicitly (maybe with catch-all patterns), moving to `Invalid_argument` to a catchable exception is a backwards-compatible change. (Here we are not talking about changing an existing exception, but my reasoning is that the inconsistency should similarly not introduce any bugs in well-written user code.)

Reply:
> If people respect our guidelines to not catch Invalid_argument explicitly (maybe with catch-all patterns), moving to Invalid_argument to a catchable exception is a backwards-compatible change.

I don't consider `Array.of_list` to be a mistake and no one asks programmers to catch `Invalid_argument`. If they don't want the exception they can check the constraint before themselves and raise or act the way they wish.

What you need to ask yourself is:

> What meaningful error strategy do you expect clients to build on raising `Failure` here ? 

The answer is: likely none, for almost 99.999% of the programs out there the program will have to be redesigned and switched to a different datastructure (and for the remaining 0.001% they still get a chance to avoid the raise). 

So the reason to raise `Invalid_argument` and not `Failure` here is that you (implicitely) made a correctness assumption on the limits of your program and they are now being violated. `Dynarray` is not able to operate within such constraints.

Of course it should also be mentioned that on OCaml's newly supported platform of choice (i.e. 64-bit computing) `Sys.max_array_length` is `18014398509481983`. If you fill that with bytes that's 18.1PB. I highly suspect something else will happen in your system before you get to see that exception. So don't bother me with an exception that I'm supposed to handle.

Reply:
I think that we are making different assumptions about the respective meaning of `Invalid_argument` and `Failure` here. For me, `Invalid_argument` "signalling a programming error" means that a correctly-written program should never ever raise it, in particular that it is the duty of the programmer calling an Invalid_argument-raising function to ensure that the error never occurs -- by a defensive check on the caller side, or by static knowledge on the callsites. On the contrary, `Failure` not being a programming error means that it is acceptable (not a bug to be fixed) if some runs of the program fail with this error, it means that the user input was outside the supported boundaries of the program.

In the case of array-creating functions, I do expect most programs to either statically ensure that the requested length is non-negative, or to add a dynamic check. On the other hand I wouldn't recommend adding a dynamic check for `Sys.max_array_length` -- in parts of the reason you mention that it never happens in practice. In *my view* of those two exceptions, this programming choice being okay (and not a programming error) is a natural consequence of negative inputs raising Invalid_argument and unrepresentable-size inputs raising Failure.

I'm not saying that my view is more correct than yours, there is a disconcerting level of subjectivity here for standard exceptions, and their documentation is very unclear on this. What I don't really understand from our present exchange is what distinction you are making between these two exceptions -- is it exactly the opposite to mine, as some parts of your answer would suggest? Something else?

Reply:
> In my view of those two exceptions, this programming choice being okay (and not a programming error) is a natural consequence of negative inputs raising Invalid_argument and unrepresentable-size inputs raising Failure.

I don't understand what you get by raising `Failure` in this case except luring the client into thinking that an error strategy should be devised – because `Failure` is meant to be caught.

Imagine we have an exception handling checking mechanism. You would likely opt-out `Invalid_argument` of it because the exception is not meant to be caught (except in `bool_of_string` of course). But you wouldn't opt-out `Failure _` because it's meant to be catched (e.g. in `int_of_string`).

Now the mechanism would constantly complain that my code using `Dynarray` is not handling `Failure` despite the fact that, as we seem to agree, it makes absolutely no sense to handle it.

> What I don't really understand from our present exchange is what distinction you are making between these two exceptions -- is it exactly the opposite to mine, as some parts of your answer would suggest? Something else?

1. `Invalid_argument` is for any condition that a function won't handle and it's the duty of the client to make sure it does not to happen. For which "won't happen in practice" is a perfectly legitimate way of "making sure it does not happen" – until hell freezes over.

2. `Failure` (or rather `Error` result) is for any condition that a function won't handle and it's the duty of the client to handle it. It is however not the client's duty to make sure it does not happen. Typical example: parsing function.

Reply:
It looks like are somehow seeing flip sides of the same coin. I insist that good clients should ensure they never raise `Invalid_argument`, and you insist that they should always handle `Failure`. (For me this is not clear/immediate, accepting to propagate certain failures to the user looks appropriate to me, not necessarily a programming error.)

In any case, I feel like I am too lazy to think *deeply* about this right now, and willing to believe (now that you clarified that yours follows a consistent logic) that both our positions have equal merit. In case of equality, we should defer to previous practice and consistency with previous modules, so your argument that `Array` follows your approach certainly holds weight. I will move to `Invalid_argument` for too-large values, although I remain secretly unconvinced.

Reply:
If `Failure` means an error/an exception the client must catch, then it would be helpful to have a clarification in the documentation or somewhere. Currently a lot of places in the runtime do not follow this semantics (grep `caml_failwith` in C code), and do not document the possibility of a `Failure` exception. (I do not have a strong opinion about the exact meaning of `Failure`, but I agree with @dbuenzli's reasoning under this assumption.)